File: codigo/vigas/Timoshenko_python/f_exacta_carga_trapezoidal_T.py
# -*- coding: utf-8 -*-
'''
Programa para encontrar las cargas nodales equivalentes asociados a una carga
trapezoidal:

   /|                                      ^  q(x)
   /|                             ^  |  |  |  |\
   /|                    ^  |  |  |  |  |  |  |\
   /|           ^  |  |  |  |  |  |  |  |  |  |\
   /|  ^  |  |  |  |  |  |  |  |  |  |  |  |  |\
   /|  |  |  |  |  |  |  |  |  |  |  |  |  |  |\
   /|  |  |  |  |  |  |  |  |  |  |  |  |  |  |\
   /|  |  |  |  |  |  |  |  |  |  |  |  |  |  |\
   /|#########################################|\
   /|                                         |\

    |--------------------L--------------------|

Para una viga de Timoshenko, asumiendo E, I, G, A constantes
'''

# %%Importación de librerías
from sympy import symbols, integrate, solve, simplify, pretty


# %%Definción de variables
x, L, w1, w2, EI, GAast = symbols('x L w1 w2 EI GAast')
w, t, M, V = symbols('w, t, M, V')
C1, C2, C3, C4 = symbols('C1, C2, C3, C4')

q = (w2-w1)*x/L + w1  # Carga trapezoidal

# %%Solución de la ecuación diferencial de Timoshenko.
V = integrate(q, x) + C1  # Ecuaciónes calculadas.
M = integrate(V, x) + C2
t = (integrate(M, x) + C3)/EI
w = integrate(t - V/GAast, x) + C4

sol = solve([w.subs(x, 0),  # Condiciones de frontera.
             w.subs(x, L),
             t.subs(x, 0),
             t.subs(x, L)],
            [C1, C2, C3, C4])

constantes = [(C1, sol[C1]), (C2, sol[C2]), (C3, sol[C3]), (C4, sol[C4])]

print(f'\nq = \n{pretty(q)}')
print(f'\nw = \n{pretty(simplify(w.subs(constantes)))}')
print(f'\nt = \n{pretty(simplify(t.subs(constantes)))}')
print(f'\nM = \n{pretty(simplify(M.subs(constantes)))}')
print(f'\nV = \n{pretty(simplify(V.subs(constantes)))}')

# %%Se evalúan las cargas nodales equivalentes
print(f'\nY1 = \n{pretty(simplify(-V.subs(constantes).subs(x, 0)))}')
print(f'\nY2 = \n{pretty(simplify(+V.subs(constantes).subs(x, L)))}')

# %%Y los momentos nodales equivalentes
print(f'\nM1 = \n{pretty(simplify(+M.subs(constantes).subs(x, 0)))}')
print(f'\nM2 = \n{pretty(simplify(-M.subs(constantes).subs(x, L)))}')

# Fin del programa

File: codigo/vigas/Timoshenko/f_exacta_carga_trapezoidal_T.m
% Programa para encontrar las cargas nodales equivalentes asociados a una carga
% trapezoidal:
%
%   /|                                      ^  q(x)
%   /|                             ^  |  |  |                   
%   /|                    ^  |  |  |  |  |  |  |\
%   /|           ^  |  |  |  |  |  |  |  |  |  |\
%   /|  ^  |  |  |  |  |  |  |  |  |  |  |  |  |\
%   /|  |  |  |  |  |  |  |  |  |  |  |  |  |  |\  
%   /|  |  |  |  |  |  |  |  |  |  |  |  |  |  |\
%   /|  |  |  |  |  |  |  |  |  |  |  |  |  |  |\
%   /|#########################################|\
%   /|                                         |\
%
%    |--------------------L--------------------|
% 
% para una viga de Timoshenko, asumiendo E, I, G, A constante 

clear, clc
syms w x L w1 w2 V(x) M(x) t(x) v(x) EI beta
GAast = (12 * EI)/(L^2 * beta);

q = (w2-w1)*x/L + w1;       % carga trapezoidal

sol = dsolve(...       
       diff(V) == q,    ... % se definen las ecuaciones diferenciales
       diff(M) == V,    ...
       diff(t) == M/EI, ... 
       diff(v) == t - V/GAast,    ...
       v(0) == 0,       ... % con sus respectivas condiciones de frontera
       v(L) == 0,       ... % para una viga doblemente empotrada
       t(0) == 0,       ...   
       t(L) == 0);
      
disp('q = '); disp(simplify(q));
disp('V = '); disp(simplify(sol.V));
disp('M = '); disp(simplify(sol.M));
disp('t = '); disp(simplify(sol.t));
disp('v = '); disp(simplify(sol.v));

% Se evaluan las cargas nodales equivalentes
Y1 = simplify(-subs(sol.V, x, 0));
Y2 = simplify(+subs(sol.V, x, L));
disp('Y1 = '); pretty(Y1);
disp('Y2 = '); pretty(Y2);

% Y los momentos nodales equivalentes
M1 = simplify(+subs(sol.M, x, 0));
M2 = simplify(-subs(sol.M, x, L));
disp('M1 = '); pretty(M1);
disp('M2 = '); pretty(M2);

%% Se imprime el vector de fuerzas nodales equivalentes
f = [Y1; M1; Y2; M2];
coef = 1/(beta + 1);
disp('f_TE = 1/(beta + 1) *')
pretty(simplify(expand(f/coef)));

%% Se compara dicho vector con la teoria de Euler-Bernoulli
disp('f_EB = ')
pretty(simplify(limit(f, beta, 0)))